src/compile/scale/component.ts

```diff
@@ -31,16 +31,37 @@ export class ScaleComponent extends Split<ScaleComponentProps> {
   }
 
   /**
-   * Whether the scale definitely includes zero in the domain
+   * Whether the scale definitely includes or not include zero in the domain
    */
-  public domainDefinitelyIncludesZero() {
-    if (this.get('zero') !== false) {
-      return true;
+  public domainHasZero(): 'definitely' | 'definitely-not' | 'maybe' {
+    const scaleType = this.get('type');
+    if (contains([ScaleType.LOG, ScaleType.TIME, ScaleType.UTC], scaleType)) {
+      // Log scales cannot have zero.
+      // Zero in time scale is arbitrary, and does not affect ratio.
+      // (Time is an interval level of measurement, not ratio).
+      // See https://en.wikipedia.org/wiki/Level_of_measurement for more info.
+      return 'definitely-not';
     }
-    return some(
-      this.get('domains'),
-      d => isArray(d) && d.length === 2 && isNumber(d[0]) && d[0] <= 0 && isNumber(d[1]) && d[1] >= 0
-    );
+
+    const scaleZero = this.get('zero');
+    if (
+      scaleZero === true ||
+      // If zero is undefined, linear/sqrt/pow scales have zero by default.
+      (scaleType === undefined && contains([ScaleType.LINEAR, ScaleType.SQRT, ScaleType.POW], scaleType))
+    ) {
+      return 'definitely';
+    }
+
+    const domains = this.get('domains');
+
+    if (domains.length > 0) {
```

Comment:
I might write
```
if (domains.length === 0) {
  return 'maybe';
}
// the rest
```

Reply:
I'd say the code here isn't long enough to use the early return pattern, which sometimes make the code harder to read, since readers need to negate the logic in their head.

To me, it's easier to follow that the code directly says that do something if the length is more than zero.  

(For longer methods, I agree that early return is good.) In any case, let's not die on this hill.